File: leetcode/Binary Tree Zigzag Level Order Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def zigzagLevelOrder(self, root):
        if root:
            queue = [root]
            zigzagOrder = []
            level = 0
            while queue:
                temp_queue = []
                for i in queue:
                    if i.left:
                        temp_queue.append(i.left)
                    if i.right:
                        temp_queue.append(i.right)
                if level % 2 != 0:
                    zigzagOrder.append([n.val for n in queue][::-1])
                else:
                    zigzagOrder.append([n.val for n in queue])
                queue = temp_queue
                level += 1
            return zigzagOrder
        else:
            return []

File: lintcode/Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
 

class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: A list of lists of integer include 
     *          the zigzag level order traversal of its nodes' values 
     */
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode *root) {
        vector<vector<int> > res;
        if (NULL == root)
            return res;
        vector<TreeNode*> queue;
        queue.push_back(root);
        
        zigzagLevelOrderHelper(queue, res, 1);
        return res;
    }
    
    void zigzagLevelOrderHelper(vector<TreeNode*> &q, vector<vector<int> > &res, int flag) {
        if (q.size()) {
            vector<TreeNode*> tmp_l;
            for (int i = 0; i < q.size(); ++i) {
                if (q[i]->left)
                    tmp_l.push_back(q[i]->left);
                if (q[i]->right)
                    tmp_l.push_back(q[i]->right);
            }
            vector<int> v;
            if (flag) {
                for (vector<TreeNode*>::iterator iter = q.begin(); iter != q.end(); ++iter) {
                    v.push_back((*iter)->val);
                }
            } else {
                for (vector<TreeNode*>::reverse_iterator iter = q.rbegin(); iter != q.rend(); ++iter) {
                    v.push_back((*iter)->val);
                }
            }
            res.push_back(v);
            q = tmp_l;
            zigzagLevelOrderHelper(q, res, !flag);
        }
    }
};
